src/aws-sdk/index.ts

```
@@ -0,0 +1,241 @@
+    );
+  }
+
+  if (!isObjectLiteralExpr(inputExpr)) {
+    throw new SynthError(
+      ErrorCodes.Expected_an_object_literal,
+      `Argument ('input') into a SDK call should be an object, found ${inputExpr.kindName}`
+    );
+  }
+
+  const params = inputExpr.getProperty("params");
+  const iamResources = inputExpr.getProperty("iamResources");
+  const iamActions = inputExpr.getProperty("iamActions");
+  const iamConditions = inputExpr.getProperty("iamConditions");
+  const aslServiceName = inputExpr.getProperty("aslServiceName");
+
+  if (!iamResources) {
+    throw new SynthError(
+      ErrorCodes.Invalid_Input,
+      `Option 'iamResources' of a SDK call is required`
+    );
+  }
+
+  if (
+    !isPropAssignExpr(iamResources) ||
+    !isArrayLiteralExpr(iamResources.expr)
+  ) {
+    throw new SynthError(
+      ErrorCodes.Invalid_Input,
+      `Option 'iamResources' of a SDK call should be an array, found ${iamResources.kindName}`
+    );
+  }
+
+  if (
+    iamActions &&
+    (!isPropAssignExpr(iamActions) || !isArrayLiteralExpr(iamActions.expr))
+  ) {
+    throw new SynthError(
+      ErrorCodes.Invalid_Input,
+      `Option 'iamActions' of a SDK call should be an array, found ${iamActions.kindName}`
+    );
+  }
+
+  if (
+    aslServiceName &&
+    (!isPropAssignExpr(aslServiceName) ||
+      !isStringLiteralExpr(aslServiceName.expr))
+  ) {
+    throw new SynthError(
+      ErrorCodes.Invalid_Input,
+      `Option 'iamActions' of a SDK call should be an array, found ${aslServiceName.kindName}`
+    );
+  }
+
+  if (
+    iamConditions &&
+    (!isPropAssignExpr(iamConditions) ||
+      // TODO: I don't think this is correct as the iamConditions most likely can be a variable as well, maybe a evalConstant here?
+      !isObjectLiteralExpr(iamConditions.expr))
+  ) {
+    throw new SynthError(
+      ErrorCodes.Expected_an_object_literal,
+      `Option 'iamConditions' of a SDK call should be an object, found ${iamConditions.kindName}`
+    );
+  }
```

For all of these, you should run `evalToConstant` and then fail when it returns undefined or an invalid runtime object type.

```ts
const iamResourcesValue = evalToConstant(iamResources.expr);
if(!iamResourcesValue || Array.isArray(iamResourcesValue) || iamResourcesValue.some(r => typeof r !== "string")) {
   ....
}
```

This will do a deep evaluation of the object to ensure all are literals or captured variables at synth time.

For example, this will work...

```ts
const arns = myConstructs.map(c => c.arn);
new StepFunction(this, 'sfn', sync () => {
   SDK.service.op(payload, { iamResources: arns });l
});
```

This works because `arns` will be a captured in a `ReferenceExpr` and during `evalToConstant` we get the captured value from the referenceExpr and return it (including applying properties/element access) to the object/arrays.

---

Why? This should never be evaluated during synth time right? The IAM resources shouldn't create states.